felix/ipsets/ipsets.go

```diff
@@ -1092,31 +1092,29 @@ func (s *IPSets) updateDirtiness(name string) {
 	}
 }
 
-func (s *IPSets) SetFilter(ipSetNames set.Set[string]) {
-	oldSetNames := s.neededIPSetNames
-	if oldSetNames == nil && ipSetNames == nil {
-		return
-	}
-	s.logCxt.Debugf("Filtering to needed IP set names: %v", ipSetNames)
-	s.neededIPSetNames = ipSetNames
-	for name, meta := range s.setNameToAllMetadata {
-		if s.ipSetNeeded(name) {
-			s.setNameToProgrammedMetadata.Desired().Set(name, meta)
-		} else {
-			s.setNameToProgrammedMetadata.Desired().Delete(name)
-		}
-		s.updateDirtiness(name)
-	}
+func (s *IPSets) SetFilter(fn func(string) bool) {
+	s.filterIPSet = fn
 }
 
 func (s *IPSets) ipSetNeeded(name string) bool {
-	if s.neededIPSetNames == nil {
+	if s.filterIPSet == nil {
 		// We're not filtering down to a "needed" set, so all IP sets are needed.
 		return true
 	}
 
 	// We are filtering down, so compare against the needed set.
-	return s.neededIPSetNames.Contains(name)
+	return s.filterIPSet(name)
+}
+
+func (s *IPSets) MarkDirty(ipsetNames set.Set[string]) {
+	for name, meta := range s.setNameToAllMetadata {
+		if ipsetNames.Contains(name) {
```

Comment:
This is now the wrong check because `s.filterIPSet` should govern whether the IP set is programmed.  You should remove the `ipsetNames` parameter and use the filter.  Then this method can be called `OnFilterUpdated` or something like that and it just triggers a recheck of the filter for all IP sets.